DataTable column controls with sane defaults to improve performance

adamconroy opened the pull request:
Feature to fix #1771 

Currently, DataTables have a severe performance issue when rendering more than a few dozen columns. The causes of this issue lie within multiple parts of the system and little short of rewriting both the front and middle ends of this code would fully solve the issue.

As a stop gap this feature allows the user full control over which columns to render with sane defaults for respectable performance in the average case. The buttons allow the user to easily navigate the visible columns and the textbox gives the user full control over the visible columns, retaining the ability to render the full table if they desire.

![ianzu94](https://user-images.githubusercontent.com/136863/48403752-94f48500-e6fc-11e8-90e2-9d52dfa5cc0a.png)
[Figure 1: Wide DataTable only rendering 50 columns at a time]

kevinushey commented:
Can we try to avoid whitespace-only changes when possible? They make the diffs noisier and hence the PRs a bit harder to review.

kevinushey commented:
Would it be worth adding a label to the column selection widget, e.g.

```
Showing columns:    <<  <  1 - 50  >  >>
```

It might also be worth using dedicated icons (or maybe there's an appropriate unicode character) in place of the plain `<` and `>` characters.

kevinushey commented:
Minor issue: if you unfocus and refocus the IDE window while the data viewer is showing, the separator between the Filter widget and the column selection widget shows up.

Before:
![screen shot 2018-11-13 at 9 26 23 am](https://user-images.githubusercontent.com/1976582/48431293-393ef180-e726-11e8-9b02-3b5079b11eb3.png)

After:
![screen shot 2018-11-13 at 9 26 27 am](https://user-images.githubusercontent.com/1976582/48431292-393ef180-e726-11e8-8886-14672370522c.png)

kevinushey commented:
I think it would also fit better if the pager were placed at the bottom right, in the same bar as the 'Showing x to y of z entries' label.

kevinushey commented:
Also worth thinking about: there will likely be users who want to view column 50 and 51 on the same page, but will be unable to with the current implementation -- e.g. for such users this is effectively a regression in functionality. Would it be worth accommodating this use case as well; e.g. by making it possible to enable / disable the pager, or specify a custom column limit so more (or fewer) columns can be viewed at once?

kevinushey commented on src/cpp/session/resources/grid/datatables/js/jquery.dataTables.js:
```diff
@@ -927,18 +927,7 @@
-						fn( aTargets[j], def );
-					}
-					else if ( typeof aTargets[j] === 'number' && aTargets[j] < 0 )
+					if ( typeof aTargets[j] === 'number' && aTargets[j] < 0 )
```
Do we have a mechanism for preserving these patches if we update jQuery in the future?

adamconroy replied:
I'm not sure, though this is an iframe'd in javascript that lives in its own sandbox and the versions of jQuery and DataTables in here are quite old. I'm of the mind right now to push for a rewrite of the DataTable feature in general when it is able to get priority at this point regardless.

jmcphers replied:
I think that when/if we rewrite the data viewer we should use another library entirely. DataTables is not really up to the task. PhosphorJS for example has a datagrid library that purportedly can handle arbitrarily large data (they have a demo that does 1 trillion rows/cols smoothly).

https://github.com/phosphorjs/phosphor/tree/master/packages/datagrid

adamconroy replied:
Absolutely. DataTables/jQuery are very, very old tech at this point make things way harder than they have to be. A Flexbox-based styling would literally cut DataTable's code in half, or less. There's a lot of stuff that it does that is just antiquated these days.

kevinushey replied:
Sounds good -- if we're anticipating an infrastructure overhaul next time we spend substantial time on the viewer then I think it'd fine to just patch directly.

kevinushey commented on src/gwt/src/org/rstudio/core/client/widget/DataTableColumnWidget.java:
```diff
@@ -0,0 +1,92 @@
+      {
+         // starting at column 1 means no offset
+         offsetAndMax[0] = Integer.parseInt(values[0].trim()) - 1;
+      } catch (NumberFormatException e)
```
Can we keep `}` and `catch()` on separate lines? At least in Java sources we almost always keep `{` and `}` on their own lines (with exceptions for e.g. lambdas)

adamconroy replied:
Yeah this is all a matter of me configuring IntelliJ's auto formatter, I must have missed this one as all of the defaults have things on the same line. It's also compounded by going between Javascript and Java using differing styles.

kevinushey commented on src/gwt/src/org/rstudio/studio/client/dataviewer/DataTable.java:
```diff
@@ -98,7 +97,98 @@ public void onValueChange(ValueChangeEvent<String> event)
+   {
+      return (offset, max) ->
+      {
+         // This is super weird but the type coming in is throwing crazy exceptions if
```
This might be because GWT's integer type is a wrapped integer as opposed to a 'native' JavaScript number.

adamconroy replied:
Ah interesting! I tried doing all sorts of debug logging to try to get to the meat of the issue but this was the only solution that ended up working for me. Would love to try something cleaner.

adamconroy commented on src/gwt/src/org/rstudio/core/client/widget/DataTableColumnWidget.java:
```diff
@@ -0,0 +1,92 @@
+package org.rstudio.core.client.widget;
+
+import com.google.gwt.user.client.ui.TextBox;
+import com.google.gwt.event.dom.client.*;
```
This was IntelliJ auto changing it to `*` because I incidentally did include everything inside the package. If I were not using everything in the package it would automatically change it to the correct includes. It's probably a setting I can change to prefer being verbose versus using `*` when it's not overly inclusive for line reduction. I'm fine with this either way.

adamconroy commented:
@kevinushey 

> Can we try to avoid whitespace-only changes when possible? They make the diffs noisier and hence the PRs a bit harder to review.

Sorry this was IntelliJ's auto formatter. What I've done in the past is view changelists with appending `?w=1` to the URL which shows you the code changes ignoring the whitespace. If this isn't good enough I can change my auto formatter. I traditionally prefer to heal code as its touched even though it makes everything slightly more annoying. I find this especially important when it comes to Java code.

> Would it be worth adding a label to the column selection widget, e.g.
> It might also be worth using dedicated icons (or maybe there's an appropriate set of unicode characters) in place of the plain < and > characters.

I had a label there...somehow it got lost probably when I was cleaning the code before opening the PR. I'll get it back in there. Unfortunately anything longer than "Cols:" takes up what little space we have there. I like the idea of using better characters. I'll see what we have for unicode, if not I'll grab FontAwesome (if we don't have it already) and use something they have. I agree the ASCII characters look a little...minimalist.

> Minor issue: if you unfocus and refocus the IDE window while the data viewer is showing, the separator between the Filter widget and the column selection widget shows up.

It should actually always be visible! I'm not sure what happened there. Might be related to the "Cols:" label as well. Good catch.

> I think it would also fit better if the pager were placed at the bottom right, in the same bar as the 'Showing x to y of z entries' label.

I agree, unfortunately that is part of the iframe and not part of the GWT chrome around it. Due to the DataTable itself being an iframe in its own sandbox there are some limitations with what we can do in the UI (as you can see by the struggle that I had just getting the max number of columns visible in the bottom bar, that took me a full day ending with me throwing in the towel and using a `window.` variable).

> I see now this is possible by clicking on the range and supplying your own; that's awesome! I think this isn't incredibly obvious, though -- is there some other way we could advertise that it's an editable field?

Talking with Jonathan we decided to go with a "editing this is more of an advanced feature that's the uncommon case" UI treatment. That said, I agree we could do something a bit more to make it pop out as an editable field. Maybe even just an onBlur effect with a subtle highlight. I'll play around with it. :)

kevinushey commented:
> I agree, unfortunately that is part of the iframe and not part of the GWT chrome around it. Due to the DataTable itself being an iframe in its own sandbox there are some limitations with what we can do in the UI (as you can see by the struggle that I had just getting the max number of columns visible in the bottom bar, that took me a full day ending with me throwing in the towel and using a window. variable).

FWIW we use that pattern fairly frequently for cross-window or cross-iframe communication (place a callback on the window object). It's gross, but it works, and since we (mostly) don't need to worry about collisions between the RStudio JS and other libraries we can usually get away with it.

> Talking with Jonathan we decided to go with a "editing this is more of an advanced feature that's the uncommon case" UI treatment. That said, I agree we could do something a bit more to make it pop out as an editable field. Maybe even just an onBlur effect with a subtle highlight. I'll play around with it. :)

There might be some kind of light hover treatment that could work here -- e.g. if you hover over the range, we show a background or similar that indicates "hey, this thing is clickable right now".

jmcphers commented:
> There might be some kind of light hover treatment that could work here -- e.g. if you hover over the range, we show a background or similar that indicates "hey, this thing is clickable right now".

+1, I think a hover treatment is what's needed here.

adamconroy commented:
[https://i.imgur.com/W6mR0qK.png](url)

Cleaned up the visuals and added a border on hover.

adamconroy commented:
@jmcphers @kevinushey I have addressed all feedback.

The one item that I did not change was the "events being dropped while bootstrapping" issue. The UX feel is adequate and sensical to me and the engineering to make it queue properly outside of the single threaded javascript would only add to the fragility of the functionality (and take 1/2 to 1 full day to write and test properly).

kevinushey commented:
Some other things that might be worth considering (would like to hear what both @adamconroy and @jmcphers think):

1. Can we make the default number of columns shown on each page configurable? Some users might find that 50 columns isn't enough.

2. What about a checkbox (or similar) for showing all columns? I think the most common action when users want to adjust the number of columns shown will just be "please show me everything". I'm not sure if we have space for such a UI affordance though.

jmcphers commented:
> What about a checkbox (or similar) for showing all columns?

A great question. This is certainly one of those things that people are going to ask for. However I think that it's going to be so tempting that it will be checked all the time! ("Why isn't RStudio showing me everything?! Oh, I see, I have to check a box every time so I can see everything. Dumb. *checks box habitually*")  

>  If requested by the community a "show all columns" button can be provided with a few hours' work.

I think @kevinushey was suggesting that users could change the default of 50, somewhere in Tools -> Global Options. That way they can adjust it based on their machine's capabilities and their personal level of patience. ;-) Easy to do and probably worth doing (there's almost no value in the IDE that people haven't wanted to change!), but not necessary to do before merging this PR.

My proposal is to do the following, probably in a follow-up PR if we have time:

a) add a default to Global Options which makes the value of 50 user-settable, and
b) allow the special string `all` as a default as well as an acceptable value in the textbox in the dataviewer, indicating that all columns should be shown.

Not at all discoverable, of course, but that's kind of the point; we don't want people falling into the configuration by accident, just a way for the really frustrated folks to get the behavior they think they want. 

@kevinushey I'm hoping to merge this by EOD so let us know if you have any other feedback!

kevinushey commented:
That's all the feedback I have :-) LGTM!

jmcphers commented:
Okay, thanks! @adamconroy congrats on your first big merge! :-)
